ringbuffer/overtime.go

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) The Thanos Community Authors.
+// Licensed under the Apache License 2.0.
+
+package ringbuffer
+
+import (
+	"context"
+	"math"
+
+	"github.com/thanos-io/promql-engine/execution/telemetry"
+	"github.com/thanos-io/promql-engine/query"
+
+	"github.com/prometheus/prometheus/model/histogram"
+)
+
+// TODO: this is mostly copied over from the rate aligner and not at all special to
+// count_over_time - we could of course generalize it to other over_time functions and we
+// will do that soon.
+// This is really just to get warmed up.
+// The idea here is that we tile the range into "stepRanges" and when we push a sample
+// and it happens to fall into a step range, we increment the state that we hold in a
+// corresponding "resultSamples" slice.
+
+// CountOverTimeBuffer is a Buffer which can calculate count_over_time for a series in a
+// streaming manner, calculating the value incrementally for each step where the sample is used.
+type CountOverTimeBuffer struct {
+	// stepRanges contain the bounds and number of samples for each evaluation step.
+	stepRanges []stepRange
+
+	// resultscounts contains the resulting state for each evaluation step.
+	resultCounts []float64
+
+	// firstTimestamps contains the timestamp of the first sample for each evaluation step.
+	firstTimestamps []int64
+
+	// lastTimestamp is the timestamp of the lsat sample in the current evaluation step
+	lastTimestamp int64
+
+	step int64
+}
+
+// NewCountOverTime creates a new CountOverTimeBuffer.
+func NewCountOverTimeBuffer(opts query.Options, selectRange, offset int64) *CountOverTimeBuffer {
+	var (
+		step     = max(1, opts.Step.Milliseconds())
+		numSteps = min(
+			(selectRange-1)/step+1,
+			querySteps(opts),
+		)
+
+		current         = opts.Start.UnixMilli()
+		resultCounts    = make([]float64, 0, numSteps)
+		firstTimestamps = make([]int64, 0, numSteps)
+		stepRanges      = make([]stepRange, 0, numSteps)
+	)
+	for range int(numSteps) {
+		var (
+			maxt = current - offset
+			mint = maxt - selectRange
+		)
+		stepRanges = append(stepRanges, stepRange{mint: mint, maxt: maxt})
+		resultCounts = append(resultCounts, 0.)
+		firstTimestamps = append(firstTimestamps, math.MaxInt64)
+
+		current += step
+	}
+
+	return &CountOverTimeBuffer{
+		step:            step,
+		stepRanges:      stepRanges,
+		resultCounts:    resultCounts,
+		firstTimestamps: firstTimestamps,
+		lastTimestamp:   math.MinInt64,
+	}
+}
+
+func (r *CountOverTimeBuffer) SampleCount() int {
+	return r.stepRanges[0].sampleCount
+}
+
+func (r *CountOverTimeBuffer) MaxT() int64 { return r.lastTimestamp }
+
+func (r *CountOverTimeBuffer) Push(t int64, v Value) {
+	// Set the lastSample sample for the current evaluation step.
+	r.lastTimestamp = t
+
+	// Set the first sample for each evaluation step where the currently read sample is used.
+	for i := 0; i < len(r.stepRanges) && t > r.stepRanges[i].mint && t <= r.stepRanges[i].maxt; i++ {
```

Comment:
For range queries we can have overlapping windows between subsequent eval steps. Are we accounting for that here?

Reply:
yeah we just add to the corresponding stepRange if the sample overlaps

Reply:
since we loop over all step ranges, we will add the same sample to the overlapping ones too